DM-23024: Allow multiple datasets to be associated with a single file on ingest

timj opened the pull request:
The most work was in changing all the code that used FileDataset.ref to instead use FileDataset.refs.

cc/ @parejkoj 

I will now check what other packages need to change because of FileDataset changes.

timj committed:
Add basic ingest test

timj committed:
Add support for ingesting multiple datasets from a single file

FileDataset now uses a list of DatasetRef and ingest iterates
over them.
Formatters can now access the associated dataId so that they
can extract a subset from a file.

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -60,10 +64,11 @@ class Formatter(metaclass=ABCMeta):
     """
 
-    def __init__(self, fileDescriptor: FileDescriptor):
+    def __init__(self, fileDescriptor: FileDescriptor, dataId: "DataCoordinate" = None):
```
If you use `from __future__ import annotations`, you can write `DataCoordinate` without the quotes.

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -77,6 +82,11 @@ def fileDescriptor(self) -> FileDescriptor:
 
+    @property
+    def dataId(self) -> Dict:
+        """DataId associated with this formatter (`dict`)"""
```
If constructor takes a `DataCoordinate`, that's what the type is going to be here, too, right?

timj replied:
Yes, I should have gone back and fixed that

TallJimbo commented on python/lsst/daf/butler/core/repoTransfers.py:
```diff
@@ -57,10 +57,11 @@ class RepoTransferFormatConfig(ConfigSubset):
-    """Registry information about the dataset (`DatasetRef`).
+    refs: List[DatasetRef]
+    """Registry information about the dataset. (`DatasetRef` or `list` of
+    `DatasetRef`).
```
This is the doc for the attribute, not the ctor arg, so I think the type is just "`list` of `DatasetRef`" here.

TallJimbo reviewed (approved):
Looks good aside from some doc/annotation type stuff.  And it seems you've duplicated `butler.py` in addition to modifying `_butler.py` (presumably when resolving a rebase conflict).

timj commented:
My rebase completely failed to notice `_butler.py` turn up. 👎

timj committed:
Change iterable to not iterate Mappings

We never really expected `iterable(dict)` to return the keys of
the dict and are not relying on this anywhere. Change the
behavior so that `iterable(dict)` returns `[dict]`.

This allows DataCoordinate to be passed to iterable.

timj committed:
Fix a deprecation warning coming from afw changes

timj committed:
Fix multi-dataset ingest

Previously we were writing a whole new file each time.

timj committed:
Only remove datasets from datastore if it is the last ref to it

timj committed:
Improve ingest tests to check for removal and URI

timj committed:
Fix stringification of DatasetRef with non-none id

timj committed:
Refactor code for checking if last ref to a file is being removed

Fixes problem in Butler.remove where we were trying to remove
everything and catching exceptions. Now we only try to remove
things that we know the datastore was given.

timj committed:
Check for dataset existence before removing it

Adjust tests such that butler.remove remembers a component.
This leads to a problem unless subsequent butler.remove
on the composite checks for dataset existence first.

TallJimbo commented on python/lsst/daf/butler/core/utils.py:
```diff
@@ -54,9 +54,10 @@ def iterable(a):
+    - iterable, but not a `str`  or Mapping -> iterate over elements
       (e.g. ``[i for i in a]``)
     - a `str` -> return single element iterable (e.g. ``[a]``)
+    - a Mapping -> return single element iterable
```
This makes me a little nervous, as I think it's hard to remember what makes a class qualify as a `Mapping`.  Could we at least make this not the default behavior?  Or maybe allow general `isinstance` exceptions via an optional kwarg, i.e.:
`iterable(x, except=collections.abc.Mapping)` (though that's not terribly readable either).

timj replied:
Do you ever want iterable to return the keys of a Mapping? The problem here is dataIds and compatbility with old YAML files.

TallJimbo replied:
That's a fair point; I think you're right that this is the only behavior we'd ever actually want for mappings - my concern is more that it's not the behavior I would have naively expected.  But it's not a big deal either way.

I do think Python mappings would be much cleaner if they weren't iterable at all and you had to explicitly say `keys()`, but of course there's nothing we can do about that.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -246,6 +246,22 @@ def getStoredItemInfo(self, ref):
+            return records
+
+        for r in records:
+            print(r["path"])
```
Debug relic.

timj replied:
I think you were looking at an old PR.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -649,14 +648,21 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
-                try:
+            # There is a difference between a concrete composite and virtual
+            # composite. In a virtual composite the datastore is never
+            # given the top level DatasetRef. In the concrete composite
```
Does this mean there is a difference in behavior at the butler level when we try to delete different kinds of composites?  Or is it just butler's responsibility rather than datastore's to iterate over components in the virtual composite case?

timj replied:
In one case butler sends the composite ref to the datastore and datastore stores its own entries because it is writing a single file. In the virtual case (I still get these the wrong way round) the butler keeps the top level composite ref to itself and only sends the components to the datastore.  Previously we were doing a "send all refs regardless" approach and mostly things were throwing exceptions because either the ref was never known to butler, or else after the first ref was removed the component refs were already deleted. Now at least the two paths are separate just as in getDirect.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -246,21 +247,22 @@ def getStoredItemInfo(self, ref):
-                for r in records]
+        Returns
+        -------
+        refs : `set` of `int`
```
I'd call this something other than "refs" (same with the method name).  Maybe "ids"?

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -660,7 +660,10 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
+                    # If a dataset was removed previously but remembered
+                    # in registry, skip the removal in the datastore.
+                    if self.registry.getDatasetLocations(r):
+                        self.datastore.remove(r)
```
Note that this attempts to remove the dataset from this particular datastore if it is present in any datastore.  I think that's okay for now, but we should open a ticket to address it in the future.  The more I think about it, the more I think we'd be better off if butler had a dict of datastores instead of having `ChainedDatastore`, so it could refer to them by name in this and other places.

timj replied:
This is fine so long as the datastores used by the registry haven't changed because of a config change. I can filter the returned list by asking for the equivalent from the datastore itself.

timj replied:
I think I've fixed it now so that it does use the datastore name before deciding whether to remove.

TallJimbo reviewed (approved):
Looks good; only minor comments, though this increases my unease with `ChainedDatastore` as a way to associate multiple `Datastores` with a butler, and I'd like to address that eventually.

timj committed:
Add Datastore.names property to get all the datastore names

This is only interesting for ChainedDatastore but it provides
a standard way of getting all the names in a list from a
datastore.

timj committed:
We are returning ids so call them ids and not refs

timj committed:
Take datastore names into account before removing components

Prior to this change any dataset stored in a datastore would
trigger the remove even if the butler had changed datastores
with reconfiguration.
